JoinColony/colonyCDapp
# [Multi-sig] feat: implement multi-sig support for funding an advanced payment

## Description

This PR adds Multi-sig support for the funding step.
Note: for some reason, the payment builder widget can get stuck in a pending state, both in the confirm details and funding step... This was happening to me on `master` too, so we should definitely investigate why it happens, but I don't think we should fix it here...

[Block ingestor PR](https://github.com/JoinColony/block-ingestor/pull/290)
[Figma link](https://www.figma.com/design/SaO0Zj6e7yGqLzjEsxqVhy/Permissions?node-id=22101-46905&node-type=frame&t=88D3UBSl2i7HmUac-0)

## Testing

The testing itself isn't going to be a lot of work, but if you spot anything being broken, give me a shout.
Test it with `amy`, `fry` and `leela` with 3 browsers so we can cover the flow properly.

1. Install the Multi-sig extension, set the global threshold to 2
![image](https://github.com/user-attachments/assets/95fb9618-6d81-455d-94b0-ec08ece39b77)
2. Give `leela` Multi-sig `Owner` permissions in `General` , `amy` the Multi-sig `Payer` in `General` and `fry` Multi-sig Payer in `Andromeda`
![image](https://github.com/user-attachments/assets/61391578-95a7-4f81-849e-9102c43079e5)
![image](https://github.com/user-attachments/assets/232648bb-1676-4ae0-a6e4-1f5659a38f03)
![image](https://github.com/user-attachments/assets/c2e237d5-4c6b-439f-a6ed-e31754b04ec0)
3. Create an advanced payment action sending out tokens to `amy` and `leela` from `General`
![image](https://github.com/user-attachments/assets/711bf044-8c71-4ce2-b0cf-8e495fef73d4)
4. Get to the `Funding` step, choose Multi-sig as the decision method. Also verify that you get a notification
![image](https://github.com/user-attachments/assets/e941bffe-e622-4e27-a28c-abfb8a0bf445)
![image](https://github.com/user-attachments/assets/0b9f30f2-55b5-40a3-9c26-61b3bacd213a)
![image](https://github.com/user-attachments/assets/ad5236b6-b2bf-4762-91b7-23139b040e94)
5. Refresh the page, verify that the widget loading isn't janky (just a skeleton should appear)
6. Let's try out rejection first, Remove your vote as `leela` and then Reject the multisig. Verify that the status pill changed to `Rejected` and that you get the funding button back. Verify that you get a rejected multisig notification.
![image](https://github.com/user-attachments/assets/3988eb52-ce80-45be-94b5-5ee9314f21c1)
![image](https://github.com/user-attachments/assets/374133cb-37f6-4ff6-b2d0-e381e20807ec)
![image](https://github.com/user-attachments/assets/393193f3-4c96-4425-9c4e-3ebe1e966091)
7. Now create a multi sig motion for funding again. Verify that you can switch between the two.
![image](https://github.com/user-attachments/assets/f8c9e39b-47b7-4d1f-b85f-de3ab0ac9093)
8. Open up the payment action as `amy` and verify that she can sign it and you can switch between actions. The first one should be open by default
![image](https://github.com/user-attachments/assets/d9ec978a-b597-4c24-b718-9ec8744f6fcf)
9. Open the action up as `fry` and verify that you can't vote on it, but you can peruse at your own leisure
![image](https://github.com/user-attachments/assets/387ac769-a85b-46d1-b082-5fc300f4f4fb)
10. Right back at it, as `amy` approve the funding step and Finalize the motion. Verify that you get notifications for approval and finalization
![image](https://github.com/user-attachments/assets/247a8908-34ab-4d44-aa7f-c38705274f83)
11. Now release as `leela`, wrap it up and verify that `amy` and `leela` got the funds
![image](https://github.com/user-attachments/assets/28e77dc8-61e7-4d4f-b410-9052d70904e2)
![image](https://github.com/user-attachments/assets/4e813ae6-7100-4771-aa2d-c6f029b0246a)
Right additionally let's see if `fry` can sign from `Andromeda`.
1. As `leela` create an advanced payment sending some CREDS to 1 recipient (we test an edge case where an advanced payment to 1 recipient is parsed as just `moveFunds` in the block-ingestor.
![image](https://github.com/user-attachments/assets/a74c5ce9-a3eb-4868-ae94-fea30dcea9f1)
2. Get to the funding step and as `fry` choose Multi-sig as the decision method (just so we verify not only the payment creator can do that)
![image](https://github.com/user-attachments/assets/e8f213cf-52c1-4aef-bd96-3ea1791be57d)
![image](https://github.com/user-attachments/assets/a08d1359-6747-4b96-957f-2f4fd8dce6c1)
3. Remove your vote as `fry` and re-approve (just a sanity check)
4. As `leela` approve this motion, finalize it and see that you got the funds
![image](https://github.com/user-attachments/assets/96f36e58-0588-4933-9703-052c64a0ee12)

### Additional tests
Since I've kinda rewrote the `getMoveFundsPermissionProofs` helper please also test
1. Move funds via permissions action still works
2. Funding an expenditure via permissions still works
3. Set the threshold to something like 10, try to fund an expenditure via Multi-sig and verify that this banner appears with the submit disabled
![image](https://github.com/user-attachments/assets/3bd251b2-6d4c-4a93-97cf-27548af2e1d1)
4. Verify that you can't fund an expenditure via multi-sig if you don't have the Funding multi-sig role in that team

#### Funding with multiple tokens:

Create and advanced payment with multiple tokens and try to fund via multi-sig.
Verify that the multi-sig was created with the proper `expenditureFunding` array by running this query:
```
getColonyAction(
    id: "the action id"
  ) {
    colonyId
    expenditureId
    expenditure {
      balances {
        amount
        tokenAddress
      }
    }
  }
```
![image](https://github.com/user-attachments/assets/febc079a-de6d-4fcb-ba35-03cac33c5937)

Copy over the expenditureId and run this query
```
getMultiSigByExpenditureId(
    expenditureId: "copied expenditureId"
  ) {
    items {
      multiSigDomainId
      isExecuted
      isRejected
      expenditureFunding {
        amount
        tokenAddress
      }
    }
  }
```
![image](https://github.com/user-attachments/assets/b848bc91-d594-482b-882d-6768b3149a35)
Verify that both tokens are in the array. Finalize the funding multisig and rerun both queries (I just did it at once) and verify that the expenditure funding matches the multisig one:
![image](https://github.com/user-attachments/assets/3b77b63a-8b24-46d8-8e1a-e7b3417ec8c7)
![image](https://github.com/user-attachments/assets/06d55e3b-72b8-4250-8e8c-d761928a3dea)


## Diffs
I've noted the block-ingestor diffs in that PR.

**New stuff** ✨

* `fundExpenditureMultiSig` saga
* `MotionWidgetSkeleton` component
* `getMoveFundsActionDomain` helper which returns the domain in which the action should happen (if we are doing a motion, we pass that through and it overrides it)

**Changes** 🏗

* `PaymentBuilderWidget` now has multi-sig support
* `MultiSigWidget` can be outside the stepper, added a `variant` prop for that
* `FundingModal` can trigger funding via multi-sig
* added `expenditureFunding` to the `ColonyMultiSig` model
* `getMoveFundsPermissionProofs` now just returns the child skill indices required and realigned `moveFunds` and `fundExpenditure` sagas
* had to realign split payment and payment builder `getPayload` helpers to not always send through the createdIn domain (we really do need to split those payloads up depending on the decision method, but let's do it when we do multisig support for creating them)


Resolves #3497
